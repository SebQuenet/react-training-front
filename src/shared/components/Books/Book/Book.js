// @flow

import React, { useState } from 'react';
import { connect } from 'react-redux';
import { makeStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import { Redirect } from 'react-router';
import { setSelectedUrlBook } from '../../../redux/actions/BookActions';

const useStyles = makeStyles({
  card: {
    minWidth: 250,
    marginBottom: 12,
    marginRight: 12,
    marginLeft: 12,
    marginTop: 12,
  },
  bullet: {
    display: 'inline-block',
    margin: '0 2px',
    transform: 'scale(0.8)',
  },
  title: {
    fontSize: 14,
  },
  pos: {
    marginBottom: 12,
  },
});

type PropTypes = {
  book: Object,
  updateSelectedUrlBook: Function,
};

/**
 * Book component
 */
const Book = ({ book, updateSelectedUrlBook }: PropTypes) => {
  const classes = useStyles();
  const [redirect, setRedirect] = useState(false);

  /**
   * handle select book
   * @param url
   */
  function handleSelectBook(url: string) {
    setRedirect(true);
    updateSelectedUrlBook(url);
  }

  if (redirect) {
    return <Redirect to="/characters" />;
  }

  return (
    <Card className={classes.card}>
      <CardContent>
        <Typography className={classes.title} color="textSecondary" gutterBottom>
          Titre du Livre
        </Typography>
        <Typography variant="h5" component="h2">
          {book.name}
        </Typography>
        <Typography className={classes.pos} color="textSecondary">
          Auteur
        </Typography>
        {book.authors.map((author, index) => (
          <Typography key={index} variant="body2" component="p">
            {author}
          </Typography>
        ))}
        <br />
        <Typography className={classes.pos} color="textSecondary">
          Ã‰diteur
        </Typography>
        <Typography variant="body2" component="p">
          {book.publisher}
        </Typography>
      </CardContent>
      <CardActions>
        <Button size="small" color="primary" onClick={() => handleSelectBook(book.url)}>
          Voir plus...
        </Button>
      </CardActions>
    </Card>
  );
};

/**
 * @private
 * @param {Object} state
 * @returns {{}}
 */
const mapStateToProps = (state: string) => ({
  ...state,
});

/**
 * @private
 */
const actions = {
  updateSelectedUrlBook: setSelectedUrlBook,
};

export default connect(mapStateToProps, actions)(Book);
