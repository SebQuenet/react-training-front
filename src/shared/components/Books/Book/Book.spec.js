import React from 'react';
import renderer from 'react-test-renderer';
import Book from './Book';
import { shallow } from '../../../../../tools/config/test/enzyme';

jest.mock('react-redux', () => ({ connect: () => c => c }));

const defaultProps = {
  book: {
    url: 'https://anapioficeandfire.com/api/books/4',
    name: 'The Hedge Knight',
    isbn: '978-0976401100',
    authors: ['George R. R. Martin'],
    numberOfPages: 164,
    publisher: 'Dabel Brothers Publishing',
    country: 'United States',
    mediaType: 'GraphicNovel',
    released: '2005-03-09T00:00:00',
    characters: [
      'https://anapioficeandfire.com/api/characters/29',
      'https://anapioficeandfire.com/api/characters/38',
      'https://anapioficeandfire.com/api/characters/40',
      'https://anapioficeandfire.com/api/characters/41',
      'https://anapioficeandfire.com/api/characters/46',
      'https://anapioficeandfire.com/api/characters/54',
    ],
    povCharacters: ['https://anapioficeandfire.com/api/characters/329'],
  },
};

describe('Book component', () => {
  it('should render correctly', () => {
    // given
    const component = renderer.create(<Book {...defaultProps} />);
    // when
    const tree = component.toJSON();
    // then
    expect(tree).toMatchSnapshot();
  });

  it('should update url book when handleSelectBook is called', () => {
    // given
    const updateSelectedUrlBook = jest.fn();
    const component = shallow(
      <Book {...defaultProps} updateSelectedUrlBook={updateSelectedUrlBook} />,
    );
    // when
    component.instance().handleSelectBook('urlBook')();
    // then
    expect(updateSelectedUrlBook).toHaveBeenCalled();
  });
});
