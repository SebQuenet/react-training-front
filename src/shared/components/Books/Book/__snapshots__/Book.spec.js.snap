// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book component should render correctly 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 MuiCard-root makeStyles-card-1 MuiPaper-rounded"
>
  <div
    className="MuiCardContent-root"
  >
    <p
      className="MuiTypography-root makeStyles-title-3 MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
    >
      Titre du Livre
    </p>
    <h2
      className="MuiTypography-root MuiTypography-h5"
    >
      The Hedge Knight
    </h2>
    <p
      className="MuiTypography-root makeStyles-pos-4 MuiTypography-body1 MuiTypography-colorTextSecondary"
    >
      Auteur
    </p>
    <p
      className="MuiTypography-root MuiTypography-body2"
    >
      George R. R. Martin
    </p>
    <br />
    <p
      className="MuiTypography-root makeStyles-pos-4 MuiTypography-body1 MuiTypography-colorTextSecondary"
    >
      Ã‰diteur
    </p>
    <p
      className="MuiTypography-root MuiTypography-body2"
    >
      Dabel Brothers Publishing
    </p>
  </div>
  <div
    className="MuiCardActions-root MuiCardActions-spacing"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Voir plus...
      </span>
    </button>
  </div>
</div>
`;
